### ISSUE TITLE
Diagnosing problem with resample_data

### ISSUE THREAD

**Issue by aaronrudkin** [DeclareDesign/fabricatr#18]
I have some test nested data -- ignore how ugly this generation function is, it was a quick hack to test something, regardless you can verify that the data itself is sane:

```
set.seed(19861108)

data_gen = function()
{
  countries = 1:200
  provinces = lapply(1:200, function(i) { 1:runif(1,30,50) })
  c_set = unlist(lapply(1:200, function(i) { rep(i, length(provinces[[i]])) }))
  d = cbind(unname(c_set), unlist(provinces))
  cities = lapply(1:nrow(d), function(i) { 1:runif(1,1,10) })
  cit_set = lapply(1:nrow(d), function(i) { d[rep(i, length(cities[[i]])), ] })
  full = cbind(do.call(rbind, cit_set), unlist(cities))
  colnames(full) = c("country", "prov", "city")
  full = as.data.frame(full)
  return(full)
}

data = data_gen()
```

Short description: there are 200 countries, each of which has 30-50 provinces, each of which has 1-10 cities.

Currently, this will cause resample_data() to fail: `resample_data(data, N=c(20, 8, 5), ID_labels=c("country", "prov", "city"))`. It will only fail if you try to resample data from all three levels, just sampling the first two levels will not fail.

At some point in the process of drawing data, bootstrap_single_level will be handed a data frame with 0 rows, and trying to sample from a data frame with 0 rows will cause it to fail.

(Mostly just opening this issue to track progress in isolating what is breaking the resample_data function -- don't necessarily need feedback on this)

**Comment by aaronrudkin** [comment]
I am not entirely convinced this function is behaving sanely at all.

Imagine a very reduced case:
```
country = c(rep(1, 5), rep(2, 5), rep(3, 5), rep(4, 5), rep(5, 5))
prov = rep(1:5, 5)
df = cbind(country, prov)
colnames(df) = c("Country", "Prov")
df = as.data.frame(df)
```

Five countries, each with 5 provinces. We have a total of 25 rows. We wish to sample 5 countries from the dataset and then sample 3 provinces from each country.

```
set.seed(19861108)
output = resample_data(df, N=c(5, 3), ID_labels=c("Country", "Prov"))
```

resample_data begins by checking the first level and doing a single level bootstrap with replacement. It will draw five countries and return every row matching those countries. It draws 4, 3, 3, 4, 2 (countries 1 and 5 do not come up, 3 and 4 each come up twice). So far, so good.

Now, the for loop begins with the level 2 bootstrap. First, the data is split by country in the new_data_list code. There are three keys in this list: 2 (5x2 data), 3 (10x2 data, each row repeated twice), 4 (10x2 data, each row repeated twice). The row names in the data verify that data looks as we would expect, country 3 has rows 11, 12, 13, 14, 15, and 11.1, 12.1, 13.1, 14.1, 15.1.

Each of these country-data sets is handed to bootstrap_single_level to do the second bootstrap (choosing n=3 provinces from each country). Country 2 draws provinces 5, 2, and 1. Now country 3 draws provinces 5, 1, and 1. Because every row in country 3 is duplicated, we get back two of province 5, two of province 1, and two of province 1. In other words, rather than doing two separate bootstraps of the provinces in country 3 (because country 3 was chosen in the country bootstrap twice), the same set of provinces is returned each time.

In other words if I ask for a bootstrap of N=5 countries, and of N=3 provinces per country, what I get is a bootstrap of N=3 provinces per country, each repeated m times, where m is the number of times the country is chosen in the bootstrap of N=5 countries.

Unless I am mistaken about the intended behaviour here, I suspect this means that every multi-level bootstrap draw is systematically wrong.

**Comment by aaronrudkin** [comment]
Rewrote this function to be recursive and not exhibit the problem behaviour. Still looking to do some speed optimizations, we've been discussing in Slack about using rbindlist versus rbind.
